components/script/dom/document.rs

```diff
@@ -1601,58 +1613,66 @@ impl Document {
-                // Substep 2.
-                script.execute();
+        loop {
+            if self.script_blocking_stylesheets_count.get() > 0 {
+                return;
             }
-            // Substep 3.
-            deferred_scripts.remove(0);
-            // Substep 4 (implicit).
-        }
-        // https://html.spec.whatwg.org/multipage/#the-end step 4.
-        self.maybe_dispatch_dom_content_loaded();
-    }
-
-    /// https://html.spec.whatwg.org/multipage/#the-end step 5 and the latter parts of
-    /// https://html.spec.whatwg.org/multipage/#prepare-a-script 20.d and 20.e.
-    pub fn process_asap_scripts(&self) {
-        // Execute the first in-order asap-executed script if it's ready, repeat as required.
-        // Re-borrowing the list for each step because it can also be borrowed under execute.
-        while self.asap_in_order_scripts_list.borrow().len() > 0 {
-            let script = Root::from_ref(&*self.asap_in_order_scripts_list.borrow()[0]);
-            if !script.is_ready_to_be_executed() {
+            if let Some((element, result)) = self.deferred_scripts.take_next_ready_to_be_executed() {
+                element.execute(result);
+            } else {
                 break;
             }
-            script.execute();
-            self.asap_in_order_scripts_list.borrow_mut().remove(0);
         }
+        if self.deferred_scripts.is_empty() {
+            // https://html.spec.whatwg.org/multipage/#the-end step 4.
+            self.maybe_dispatch_dom_content_loaded();
+        }
+    }
 
-        let mut idx = 0;
-        // Re-borrowing the set for each step because it can also be borrowed under execute.
-        while idx < self.asap_scripts_set.borrow().len() {
-            let script = Root::from_ref(&*self.asap_scripts_set.borrow()[idx]);
-            if !script.is_ready_to_be_executed() {
-                idx += 1;
-                continue;
-            }
-            script.execute();
-            self.asap_scripts_set.borrow_mut().swap_remove(idx);
+    /// https://html.spec.whatwg.org/multipage/#the-end step 3.
+    /// https://html.spec.whatwg.org/multipage/#prepare-a-script step 22.d.
+    pub fn asap_script_loaded(&self, element: &HTMLScriptElement, result: ScriptResult) {
+        let mut scripts = self.asap_scripts_set.borrow_mut();
+        let idx = scripts.iter().position(|entry| &*entry.element == element).unwrap();
+        scripts.swap(0, idx);
+        scripts[0].loaded(result);
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#the-end step 3.
+    /// https://html.spec.whatwg.org/multipage/#prepare-a-script step 22.c.
+    pub fn asap_in_order_script_loaded(&self,
+                                       element: &HTMLScriptElement,
+                                       result: ScriptResult) {
+        self.asap_in_order_scripts_list.loaded(element, result);
+    }
+
+    fn process_asap_scripts(&self) {
+        let mut scripts = self.asap_scripts_set.borrow_mut();
```

Comment:
I wonder if it's possible to make this panic by doing something along the lines of
```html
<script async>
var s = document.createElement("script");
s.setAttribute("async", "");
s.src = ...;
document.appendChild(s);
</script>
```
Could you try breaking it?